doc/tracing.md

```diff
@@ -211,6 +211,58 @@ Arguments passed:
 4. The expected transaction fee as an `int64`
 5. The position of the change output as an `int32`
 
+### Context `mempool`
+
+#### Tracepoint `mempool:added`
+
+Is called when a transaction is added to the node's mempool. Passes information
+about the transaction.
+
+Arguments passed:
+1. Transaction ID (hash) as `pointer to unsigned chars` (i.e. 32 bytes in little-endian)
+2. Transaction virtual size as `uint64`
+3. Transaction fee as `int64`
```

Comment:
Would it make sense to convey information about the peer (even if just ID) as well? One of the things that I think is potentially valuable e.g. [in bmon](https://github.com/chaincodelabs/bmon/blob/master/bmon/logparse.py#L245) is tracking which peers fed you which transactions.

Reply:
I think it makes sense.

However, peer information is no longer available in functions that handle low-level mempool operations, so I propose adding a separate `mempool:accepted` tracepoint for this in `PeerManagerImpl::ProcessMessage` [here](https://github.com/bitcoin/bitcoin/blob/master/src/net_processing.cpp#L4042) (which is right where the debug log entries for transactions accepted into the memory pool are created).

The rationale for adding a separate `mempool:accepted` tracepoint instead of moving the existing `mempool:added` one from `CTxMemPool::addUnchecked` is to ensure consistency of added and removed transactions.

Details: The tracepoint call in `ProcessMessage` would depend on the success of `ProcessTransaction` [here](https://github.com/bitcoin/bitcoin/blob/master/src/net_processing.cpp#L4025). There's an edge case where a transaction added to the mempool in `MemPoolAccept::Finalize` [here](https://github.com/bitcoin/bitcoin/blob/master/src/validation.cpp#L1074) is immediately removed again seven lines below during the call to `LimitMempoolSize` and `MemPoolAccept::Finalize` will return `state.Invalid(TxValidationResult::TX_MEMPOOL_POLICY, "mempool full")`. As expected, in this case `ProcessMessage` will not call the tracepoint (or create a debug log entry). `LimitMempoolSize`, however, will at some point call `CTxMemPool::removeUnchecked`, which calls the `mempool:removed` tracepoint, resulting in tracepoint data containing a "transaction removed" event without a preceding "transaction added" event.

Please note that this problem does not arise due to poor placement of the "mempool:removed" tracepoint [here](https://github.com/virtu/bitcoin/blob/add_mempool_tracepoints/src/txmempool.cpp#L481-L488). The tracepoint is positioned right after `GetMainSignals().TransactionRemovedFromMempool`, which leads me to believe that the aforementioned consistency issue currently affects zeromq and the debug log.

Reply:
Note that adding transactions we receive over the network is only one option out of several to add something to our mempool. Can also add transactions via RPC, `loadMempool()`, wallet, etc..   

I agree that it could be useful to have information about peers that send us transactions we end up accepting.  Could call a tracepoint something like `net:tx_accepted`. Not sure if that's in-scope for this PR though.

Reply:
Couldn't this be done by using both `net:inbound_message` and `mempool:added` at the same time? That is, correlate the txids from both tracepoints to deduce which peer sent the tx.